### DISCUSSION TITLE
MessagePassing.message

### DISCUSSION THREAD

**Discussion by Chrisable** [pyg-team/pytorch_geometric#5480] [General]
for code:
```
import torch
from torch.nn import Linear, Parameter
from torch_geometric.nn import MessagePassing
from torch_geometric.utils import add_self_loops, degree

class GCNConv(MessagePassing):
    def __init__(self, in_channels, out_channels):
        super().__init__(aggr='add')  # "Add" aggregation (Step 5).
        self.lin = Linear(in_channels, out_channels, bias=False)
        self.bias = Parameter(torch.Tensor(out_channels))

        self.reset_parameters()

    def reset_parameters(self):
        self.lin.reset_parameters()
        self.bias.data.zero_()

    def forward(self, x, edge_index):
        # x has shape [N, in_channels]
        # edge_index has shape [2, E]

        # Step 1: Add self-loops to the adjacency matrix.
        edge_index, _ = add_self_loops(edge_index, num_nodes=x.size(0))

        # Step 2: Linearly transform node feature matrix.
        x = self.lin(x)

        # Step 3: Compute normalization.
        row, col = edge_index
        deg = degree(col, x.size(0), dtype=x.dtype)
        deg_inv_sqrt = deg.pow(-0.5)
        deg_inv_sqrt[deg_inv_sqrt == float('inf')] = 0
        norm = deg_inv_sqrt[row] * deg_inv_sqrt[col]

        # Step 4-5: Start propagating messages.
        out = self.propagate(edge_index, x=x, norm=norm)

        # Step 6: Apply a final bias vector.
        out += self.bias

        return out

    def message(self, x_j, norm):
        # x_j has shape [E, out_channels]

        # Step 4: Normalize node features.
        return norm.view(-1, 1) * x_j
```
What functions does the message function and the propagate function realize? Can you use a formula to express them？

**Comment by Kevin-shihello-world** [comment]
well, I'm still an undergraduate student but I think I can give a decent answer hope you like it.
The question you want to ask may be because it has a propagate function in
```python
out = self.propagate(edge_index, x=x, norm=norm)
```
while did not appear where it came from. And you do not know where the message function is used.
So remember it is an override of MessagePassing go to see it in torch_geometric.nn and you may get it that the propagate function is inherited from the MessagePassing class and it used the message function.
And also there's an important thing I think you may want to know that though it wrote GCNCONV it in fact in PYG using a messagepassing way instead of a spectral way to do with information on graph.

**Reply by rusty1s** [reply]
Yes, this is correct. The call to `self.propagate` will internally call `message` and is responsible for creating edge-level feature tensors such as `x_j` or `x_i`. Afterwards, it will aggregate all messages together according to `aggr` and return its output.
